File: Sorting_Searching/bubble_sort.c
#include <stdio.h>
#include <conio.h>
#define size 100

void bubble_sort(int a[],int);


int main()
{
	int arr[size], i, n;
	printf("\n Enter the number of elements in the array : ");
	scanf("%d", &n);
	printf("\n Enter the elements of the array: ");
	for(i=0;i<n;i++)
	{
		scanf("%d", &arr[i]);
	}
	bubble_sort(arr, n);
	printf("\n The sorted array is: \n");
	for(i=0;i<n;i++)
		printf(" %d\t", arr[i]);
	return 0;
}

void bubble_sort(int data[], int N)
{
	int i,j,k,temp;
	for(i=0; i<N ;i++){
		for(j=0;j<N-i-1;j++){
			if(data[j]>data[j+1]){
				temp=data[j];
				data[j]=data[j+1];
				data[j+1]=temp;
			}
		
		}
		printf("\n\n\n The elements of the array after %d iteration are \n", i+1);
		for(k=0;k<N;k++)
		{
			printf("%d\t ", data[k]);
		}
	}
	
}

File: src/java_codes/sortingsearching/BubbleSort.java
package java_codes.sortingsearching;

import java.util.Arrays;
import java.util.Scanner;

public final class BubbleSort {

  // Private constructor to prevent instantiation of this utility class.
  private BubbleSort() {}

  /**
   * Sorts an array in ascending order using the Bubble Sort algorithm. This implementation includes
   * an optimization to terminate early if the array becomes sorted before all passes are complete.
   *
   * @param arr The array to be sorted.
   */
  public static void sort(int[] arr) {
    if (arr == null || arr.length == 0) {
      return;
    }
    int n = arr.length;
    boolean swapped;
    for (int i = 0; i < n - 1; i++) {
      swapped = false;
      for (int j = 0; j < n - i - 1; j++) {
        if (arr[j] > arr[j + 1]) {
          // Swap arr[j] and arr[j+1]
          int temp = arr[j];
          arr[j] = arr[j + 1];
          arr[j + 1] = temp;
          swapped = true;
        }
      }
      // If no two elements were swapped by inner loop, then break
      if (!swapped) {
        break;
      }
    }
  }

  public static void main(String[] args) {
    Scanner sc = new Scanner(System.in);
    System.out.print("Enter the number of elements: ");
    int n = sc.nextInt();
    int[] arr = new int[n];
    System.out.println("Enter the elements:");
    for (int i = 0; i < n; i++) {
      arr[i] = sc.nextInt();
    }

    System.out.println("Original array: " + Arrays.toString(arr));
    sort(arr);
    System.out.println("Sorted array:   " + Arrays.toString(arr));

    sc.close();
  }
}
